[WIP - Don't review] net: lib: nrf_provisioning: Make API more flexible (#22408)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/nrf_provisioning.rst

To enable the library, set the :kconfig:option:`CONFIG_NRF_CLOUD` and :kconfig:option:`CONFIG_NRF_PROVISIONING` Kconfig options.

* :kconfig:option:`CONFIG_NRF_PROVISIONING_AUTO_START_ON_INIT` - Starts and performs provisioning on library initialization.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` - Activates scheduled provisioning at a given interval.
  If disabled, the client will only provision when requested manually using the :c:func:`nrf_provisioning_trigger_manually` function.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_VALID_DATE_TIME_TIMEOUT_SECONDS` - Time to wait for valid date-time before starting provisioning.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_INTERVAL_S` - Maximum provisioning interval, in seconds, only valid if :kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is enabled.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_SPREAD_S` - Provisioning time spread in seconds, only valid if :kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is enabled.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_INITIAL_BACKOFF` - Initial time for exponential backoff in seconds, only valid if :kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is enabled.

> * :kconfig:option:`CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG` - Root CA security tag for the Provisioning Service

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_NRF_PROVISIONING_ROOT_CA_SEC_TAG` - Root CA security tag for the Provisioning Service.
```

* :kconfig:option:`CONFIG_NRF_PROVISIONING_SAVE_CMD_ID` - Saves the latest command ID to storage.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_CUSTOM_AT` - Activates custom AT commands.

> * :kconfig:option:`CONFIG_NRF_PROVISIONING_WITH_CERT` - Provisions the root CA certificate to the security tag if the tag is empty
> * :kconfig:option:`CONFIG_NRF_PROVISIONING_RX_BUF_SZ` - Response buffer size
> * :kconfig:option:`CONFIG_NRF_PROVISIONING_TX_BUF_SZ` - Request buffer size

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_NRF_PROVISIONING_WITH_CERT` - Provisions the root CA certificate to the security tag if the tag is empty.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_RX_BUF_SZ` - Response buffer size.
* :kconfig:option:`CONFIG_NRF_PROVISIONING_TX_BUF_SZ` - Request buffer size.
```

Initialization
==============

To use the library, you must initialize it.

> This is done by calling the :c:func:`nrf_provisioning_init` function and passing the following event callback handler to receive events from the library:

Reviewer 2: ```suggestion
Call the :c:func:`nrf_provisioning_init` function and pass the following event callback handler to receive events from the library:
```

.. code-block:: c

   static void nrf_provisioning_callback(const struct nrf_provisioning_callback_data *event)
   {
      /* Handle events received from the library here */
   }

   /* Initialize the provisioning client */
   ret = nrf_provisioning_init(nrf_provisioning_callback);
   if (ret) {
       LOG_ERR("Failed to initialize provisioning client: %d", ret);
       return ret;
   }

> Once initialized, provisioning can occur in one of two ways:

Reviewer 2: ```suggestion
Once initialized, provisioning can take place in one of the following two ways:
```

* Automatically, at a configured interval set by the :kconfig:option:`CONFIG_NRF_PROVISIONING_INTERVAL_S` Kconfig option if :Kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is enabled.
  Provisioning can be triggered asynchronously using the function :c:func:`nrf_provisioning_trigger_manually`.
* Manually, by calling the :c:func:`nrf_provisioning_trigger_manually` function from the application if :Kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is disabled.

For an example of how to use the library, see the :ref:`nrf_provisioning_sample` sample.

Provisioning
============

> During provisioning the device receives a set of commands from the server.

Reviewer 2: ```suggestion
During provisioning, the device receives a set of commands from the server.
```

> This requires the device to deactivate LTE (putting the modem into offline mode) to write the commands to the modem's non-volatile memory.

Reviewer 2: ```suggestion
This requires the device to deactivate LTE (putting the modem into offline mode) to be able to write the commands to the modem's non-volatile memory.
```

This is because the modem cannot be connected while any data is being written to its storage area.

> When the option :kconfig:option:`CONFIG_NRF_PROVISIONING_AUTO_START_ON_INIT` is set, the library initializes and then starts provisioning according to the configured interval.

Reviewer 2: ```suggestion
When the  :kconfig:option:`CONFIG_NRF_PROVISIONING_AUTO_START_ON_INIT` Kconfig is set, the library initializes and starts provisioning according to the configured interval.
```

   ret = nrf_provisioning_init(nrf_provisioning_callback);
   if (ret) {
       LOG_ERR("Failed to initialize provisioning client: %d", ret);
       return ret;
   }

Once initialized, provisioning can take place in one of the following two ways:

* Automatically, at a configured interval set by the :kconfig:option:`CONFIG_NRF_PROVISIONING_INTERVAL_S` Kconfig option if :Kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is enabled.
  You can trigger provisioning asynchronously using the function :c:func:`nrf_provisioning_trigger_manually`.
* Manually, by calling the :c:func:`nrf_provisioning_trigger_manually` function from the application if :Kconfig:option:`CONFIG_NRF_PROVISIONING_SCHEDULED` is disabled.

For an example of how to use the library, see the :ref:`nrf_provisioning_sample` sample.

Provisioning
============

During provisioning, the device receives a set of commands from the server.
This requires the device to deactivate LTE (putting the modem into offline mode) to be able to write the commands to the modem's non-volatile memory.
This is because the modem cannot be connected while any data is being written to its storage area.

> When the :kconfig:option:`CONFIG_NRF_PROVISIONING_AUTO_START_ON_INIT` Kconfig is set, the library initializes and starts provisioning according to the configured interval.

Reviewer: ```suggestion
When the :kconfig:option:`CONFIG_NRF_PROVISIONING_AUTO_START_ON_INIT` Kconfig option is set, the library initializes and starts provisioning according to the configured interval.
```
